File: Projeto_8/src/infrastructure/SignalHandler.cpp
#include "SignalHandler.h"
#include <iostream>
#include <fstream>
#include <cstdlib>

void SignalHandler::inicializar() {
    // Registra os hooks de sinal do Sistema Operacional
    std::signal(SIGINT, SignalHandler::tratador);
    std::signal(SIGTERM, SignalHandler::tratador);
}

void SignalHandler::tratador(int sinal) {
    std::cerr << "\n[CRITICAL ALERT] Sinal de interrupcao recebido (" << sinal << ")." << std::endl;
    std::cerr << "[SYSTEM] Iniciando protocolo de graceful shutdown..." << std::endl;

    // Gravação de log emergencial garantindo fechamento de recursos
    std::ofstream logEmergencia("emergency_shutdown.log", std::ios::app);
    if (logEmergencia.is_open()) {
        logEmergencia << "Desligamento forcado via sinal " << sinal << ". Arquivos pendentes encerrados.\n";
        logEmergencia.flush();
        logEmergencia.close();
    }

    std::cerr << "[SYSTEM] Flush dos buffers concluido. Fechando o subsistema." << std::endl;
    std::exit(sinal);
}

File: Projeto_8/src/infrastructure/SignalHandler.py
import signal
import sys

class SignalHandler:
    @staticmethod
    def inicializar() -> None:
        """Registra os hooks do Sistema Operacional."""
        signal.signal(signal.SIGINT, SignalHandler.__tratador)
        signal.signal(signal.SIGTERM, SignalHandler.__tratador)

    @staticmethod
    def __tratador(sinal: int, frame) -> None:
        """Método estático e privado invocado pelo SO durante uma interrupção."""
        print(f"\n[CRITICAL WARNING] Sinal do SO interceptado (ID: {sinal}).")
        print("[LOG EMERGÊNCIAL] Fechando arquivos e executando flush de buffers...")
        
        # Simula o encerramento seguro
        print("[ENGINE] Sistema encerrado de forma limpa. Nenhum dado corrompido.")
        sys.exit(sinal)